boot/bootutil/include/bootutil/fault_injection_hardening.h

```diff
@@ -0,0 +1,339 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2020 Arm Limited
+ */
+
+#ifndef __FAULT_INJECTION_HARDENING_H__
+#define __FAULT_INJECTION_HARDENING_H__
+
+/* Fault injection mitigation library.
+ *
+ * Has support for different measures, which can either be enabled/disabled
+ * separately or by defining one of the MCUBOOT_FIH_PROFILEs.
+ *
+ * NOTE: These constructs against fault injection attacks are not guaranteed to
+ *       be secure for all compilers, but execution is going to be correct and
+ *       including them will certainly help to harden the code.
+ *
+ * DOUBLE_VARS makes critical variables into a tuple (x, x ^ msk). Then the
+ * correctness of x can be checked by XORing the two tuple values together. This
+ * also means that comparisons between fih_ints can be verified by doing
+ * x == y && x_msk == y_msk.
+ *
+ * GLOBAL_FAIL makes all while(1) failure loops redirect to a global failure
+ * loop. This loop has mitigations against loop escapes / unlooping. This also
+ * means that any unlooping won't immediately continue executing the function
+ * that was executing before the failure.
+ *
+ * CFI (Control Flow Integrity) creates a global counter that is incremented
+ * before every FIH_CALL of vulnerable functions. On the function return the
+ * counter is decremented, and after the return it is verified that the counter
+ * has the same value as before this process. This can be used to verify that
+ * the function has actually been called.
+ *
```

Comment:
There should be some description here that describes how this only provides a specific subset of CFI. This is checking that a specific sequence of calls are made within the same block and that the adversary is someone performing glitching. It does not cover other types of CFI such as ROP protection or JOP protection.

Reply:
Fixed